### ISSUE TITLE
[ProjectTracking]: Stateless validation StatelessNet

### ISSUE THREAD

**Issue by walnut-the-cat** [near/near-one-project-tracking#20]
# Goals
* Demonstrate stateless validation to public so they have better understanding on what we are working on
* Leverage community power to identify critical issues and stress test the network with stateless validation
* Figure out detailed MainNet launch requirement and plan based on findings

# Links to external documentations and discussions
* StatelessNet planning doc: [LINK](https://docs.google.com/document/d/13QHL23KsFblZOIVhRHKaT5UhI9XCSclx6CCqoHsGD7s/edit?usp=sharing)
* NEP: [LINK](https://github.com/near/NEPs/pull/509/)

# Estimated effort
* StatelessNet launch preparation: 2-3 weeks
* StatelessNet duration: 8+ weeks 
* StatelessNet data analysis/post-mortem: 2 weeks
* [Out of scope] Implementation of the follow up items identified during StatelessNet

# Timeline
Start StatelessNet stage 1: Late January/Early February
We aim to open StatelessNet to the public on January 29th so community members can test features and generate traffic. 
Requirements
3 weeks before
[StatelessNet plan public announcement ](https://docs.google.com/document/d/13QHL23KsFblZOIVhRHKaT5UhI9XCSclx6CCqoHsGD7s/edit#heading=h.6tpfx5fqfp0u)should be made for visibility via multiple channels including X, Telegram, and Near.org.
2 weeks before
Stateless validation MVP should go through basic tests (unit tests, integration tests, load tests).
1 week before
Stateless validation MVP must be merged into master and be ready for distribution.
StatelessNet network should be set up along with necessary monitoring tools.
EthDenver: Feb 29th - March 3rd, 2024
[Bowen Wang](mailto:bowen@near.org) plans to share findings from Stateless validation StatelessNet during EthDenver. 
Requirements
1 week before
Data from StatelessNet should be gathered and analyzed in presentable format.
2 weeks before
At least 1-2 rounds of performance optimization and stability enhancement should be in place based on incoming data.
Start StatelessNet Stage 2: Early March (at least 4 weeks after stage 1)
Expand scenario coverage of StatelessNet to start testing corner cases.
Mid-point check-in: March 15th, 2024
Continue collecting data and make improvements. 
Complete StatelessNet: March 31st, 2024 (or TBD)
Wrap up the StatelessNet test and share the final report. 
Need to have enough confidence on resilience, scalability, reliability of the network before completing StatelessNet and the end date may move around.

# Assumptions
* 

# Pre-requisites

What are the required things that need to be in place before this project can start?

# Out of scope

What works are not part of this project and will not be tackled?

**Comment by Longarithm** [comment]
### Feb 8th Core team update

* Set up first branch, node configs and simple release process for statelessnet
* Solving issues
	* https://github.com/near/nearcore/pull/10535
	* https://github.com/near/nearcore/issues/10560
* Started benchmarking
* Monitoring tools
	* Technical dashboard https://nearinc.grafana.net/d/edbdp75fzrklcb/stateless-validation?orgId=1&var-chain_id=statelessnet
	* Debug UI https://github.com/near/nearcore/pull/10524

**Comment by walnut-the-cat** [comment]
## Feb 26th update
* Majority of issues discovered at the beginning of stake wars program have been resolved now
* Orphan state pool implementation is now merged. The team will re-enable mainnet traffic mirroring soon
* [WIP] Testloop refactoring is in progress; the work is necessary to enable in-memory trie with proper testing
* [WIP] Proper genesis update and removal of mock epoch manager
* [WIP] Performance refactoring iteration based on Bowen's latest feedback. Plan to share the result by Feb 27th.
* The team started reviewing incoming reward program proposals to prepare for start of the program with Stage 2
* Community validator onboarding in progress
* Other MainNet release related topics will now be tracked in the tracking issue [46](https://github.com/near/near-one-project-tracking/issues/46)

**Comment by bowenwang1996** [comment]
> [WIP] Testloop refactoring is in progress; the work is necessary to enable in-memory trie with proper testing

How much time is this going to take?

> Proper genesis handling & test fixes

What does this mean?

**Comment by walnut-the-cat** [comment]
Regarding testloop, it's being discussed in this [Zulip thread](https://near.zulipchat.com/#narrow/stream/407237-pagoda.2Fcore.2Fstateless-validation/topic/TestLoop.20progress/near/423539311). I expect it to be mostly done this week, but am double checking with Robin.

**Comment by walnut-the-cat** [comment]
> Proper genesis handling & test fixes

This is part of @pugachAG 's Better Engineering effort to make code base more manageable and easy to develop. So far, these two PRs are related: 
* https://github.com/near/nearcore/pull/10633
* https://github.com/near/nearcore/issues/10502

It will be followed by removal of mock epoch manager

**Comment by wacban** [comment]
# March 11th Update

* discovered and debugging a number of issues 
  * stack overflow - almost fixed
  * one shard stalling under mainnet load - under investigation
  * crash in memtrie - investigation will start soon
  * empty blocks and missing chunks - looking for an owner
* the next milestone, in-memory-trie and state sync integration - implementation is in progress

**Comment by wacban** [comment]
# Aprli 5th Update

We deployed a number of new features to statelessnet but unfortunately the chain suffered tremendously due to bugs. We decided to shut down the chain as maintaining it on unstable code was consuming too much developers' time. We are going to continue development and testing on dedicated test networks. We may resume statelessnet once we've gained more confidence in it. 

* finished features
  * single shard tracking using state sync
  * memtrie loading and unloading during state sync
  * soft state witness size limit 
* fixed bugs
  * missing trie value issue
  * missing storage recording issue
  * (statelessnet only) reward calculation panic due to arithmetic overflows
  * (statelessnet only) maintenance scripts accidentaly unstaking all tokens
* misc
  * Attempted mirroring mainnet traffic but failed due to networking constraints. Search for a solution is in progress.

**Comment by walnut-the-cat** [comment]
Moving the task to 'Ready to be prioritized' from 'WIP' as we are pausing StatelessNet
